server/lib/src/idm/ldap.rs

```diff
@@ -477,6 +478,94 @@ impl LdapServer {
+        _idms: &IdmServer,
+        cr: &CompareRequest,
+        uat: &LdapBoundToken,
+        _source: Source,
+    ) -> Result<Vec<LdapMsg>, OperationError> {
+        admin_info!("Attempt LDAP CompareRequest for {}", uat.spn);
+
+        // convert to a search request, and then do the search.
+        let sr = SearchRequest {
+            msgid: cr.msgid,
+            base: cr.entry.clone(),
+            scope: LdapSearchScope::Base,
+            filter: LdapFilter::Equality(cr.atype.clone(), cr.val.clone()),
+            attrs: vec!["1.1".to_string()], // we only need to know if it exists.
+        };
+
+        trace!("Converting to Search: {:?}", sr);
+
+        fn compare_result_msg(cr: &CompareRequest, code: LdapResultCode) -> LdapMsg {
+            LdapMsg {
+                msgid: cr.msgid,
+                op: LdapOp::CompareResult(LdapResult {
+                    code,
+                    matcheddn: "".to_string(),
+                    message: "".to_string(),
+                    referral: vec![],
+                }),
+                ctrl: vec![],
+            }
+        }
+
+        let res = match self.do_search(_idms, &sr, uat, Source::Internal).await {
+            Ok(r) => r,
+            Err(OperationError::InvalidAttributeName(s)) => {
+                request_error!(?s, "Invalid Attribute Name in CompareRequest");
+
+                return Ok(vec![compare_result_msg(
+                    cr,
+                    LdapResultCode::UndefinedAttributeType,
+                )]);
+            }
+            Err(e) => {
+                request_error!(err = ?e, "Error doing search for compare request");
+
+                return Err(e);
+            }
+        };
+
+        if let Some((srd, results)) = &res.split_last() {
+            if let LdapOp::SearchResultDone(LdapResult {
+                code: LdapResultCode::Success,
+                ..
+            }) = &srd.op
+            {
+                if !results.is_empty() {
+                    return Ok(vec![compare_result_msg(cr, LdapResultCode::CompareTrue)]);
+                }
+
+                // search again to determine if entry exists at all
+                let mut sr = sr;
+                sr.filter = LdapFilter::Present(Attribute::Spn.to_string());
+
+                let res = self.do_search(_idms, &sr, uat, Source::Internal).await?;
+
+                if let Some((srd, results)) = &res.split_last() {
+                    if let LdapOp::SearchResultDone(LdapResult {
+                        code: LdapResultCode::Success,
+                        ..
+                    }) = &srd.op
+                    {
+                        return Ok(vec![compare_result_msg(
+                            cr,
+                            match results.len() {
+                                0 => LdapResultCode::NoSuchObject, // entry really does not exist
+                                _ => LdapResultCode::CompareFalse, // entry exists, but condition is not met
```

Comment:
Well it could also be that the caller lacks permissions to perform the search.

Reply:
I am not sure how this should be handled - can you lead me into the right direction, please?